fabric-mining-levels-v0 -> fabric-tool-attributes-v1

LemmaEOF opened the pull request:
Not sure if this is worth a V1 or not. I've got a bit of a gut feeling that it isn't.

Currently, the API is set up so that the mining level and mining speed of a given item is hardcoded on creation. This is unhelpful to mods like Materialisation or other tinkers-likes, which need to be able to change their mining level and speed based on NBT. This PR adds a new interface to do so, and leaves existing behavior intact otherwise. As mining level and speed are the only aspects of a tool that are affected by Fabric API, the interface doesn't have anything else at this time. In the future, it can be upgraded to add more if needed.

Commit:
Update ToolManager.java

LemmaEOF committed:
add support for dynamic mining level and speed on tools

LemmaEOF committed:
clean up some spaces

modmuss50 commented on fabric-mining-levels-v0/src/main/java/net/fabricmc/fabric/mixin/tools/MixinItemStack.java:
```diff
@@ -46,7 +47,10 @@ public void getBlockBreakingSpeed(BlockState state, CallbackInfoReturnable<Float
 			if (triState != TriState.DEFAULT) {
-				info.setReturnValue(triState.get() ? ((MiningToolItemAccessor) this.getItem()).getMiningSpeed() : 1.0F);
+				float miningSpeed;
+				if (this.getItem() instanceof DynamicTool) miningSpeed = ((DynamicTool) this.getItem()).getMiningSpeed((ItemStack)(Object) this);
```
I would suggest adding braces around this if statement, I dont know if its just me but its kinda hard to follow, there is no reason to compress it up like this.

modmuss50 commented:
Other than that, I think this looks allright, no need for v1 here, you do need to bump the version number however (in the modules build.gradle.

modmuss50 commented:
Are there any suggestions on a better interface name? `DynamicTool` doesnt really exlpain what it does very well.

LemmaEOF commented:
Yeah, I'm definitely not the best at naming things. Other suggestions are appreciated!

LemmaEOF committed:
fix up code style, slip in shears tag

LemmaEOF commented:
While I'm working on the module, I added in an updated form of #100 since it was just one line and one JSON.

Sollace commented on fabric-mining-levels-v0/src/main/java/net/fabricmc/fabric/mixin/tools/MixinItemStack.java:
```diff
@@ -48,8 +48,12 @@ public void getBlockBreakingSpeed(BlockState state, CallbackInfoReturnable<Float
+				if (this.getItem() instanceof DynamicTool) {
+					miningSpeed = ((DynamicTool) this.getItem()).getMiningSpeed((ItemStack)(Object) this);
+				}
+				else {
```
Ouch! My back!

Did you intend to use a fractured-else here?

RedstoneParadox replied:
Fractured-else?

LemmaEOF replied:
Oh no I didn't lol, I was adding braces like Modmuss suggested and forgot to put the else on the same line as the if closing. I need coffee.

Sollace replied:
> Fractured-else?

It's my name for when the style changes mid-else. I see it a lot at work. Some people actually prefer it, believe it or not.

LemmaEOF committed:
clean up fractured else

RedstoneParadox commented:
Perhaps `DynamicNbtTool`? That way it at least has `Nbt` in the name so it indicates what it's supposed to be used for.

Edit: Or just `NbtTool`.

Sollace commented on fabric-mining-levels-v0/src/main/java/net/fabricmc/fabric/api/tools/DynamicTool.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * Interface for tools whose properties are affected by NBT. Goes on the item.
+ */
+public interface DynamicTool {
```
You know, there's nothing in this interface that implies the properties _have_ to come from NBT. The implementer can literally decide to determine the values however they want.

LemmaEOF replied:
That is a fair point. I'll update the javadocs when I have the time.

Sollace commented on fabric-mining-levels-v0/src/main/java/net/fabricmc/fabric/impl/tools/ToolManager.java:
```diff
@@ -87,7 +88,9 @@ public static void registerBreakByTool(Block block, Tag<Item> tag, int miningLev
 	private static int getMiningLevel(ItemStack stack) {
-		if (stack.getItem() instanceof ToolItem) {
+		if (stack.getItem() instanceof DynamicTool) {
+			return ((DynamicTool) stack.getItem()).getMiningLevel(stack);
```
Also since the interface is literally just another version of `ToolItem` (below) I'd be inclined to call it `DynamicToolItem`.


Though it's not a class... So maybe `MiningLevelDeterminent` or something, since it doesn't _really_ have to be a tool, either, does it? It's just as valid for a block to implement the interface to be used elsewhere.

LemmaEOF replied:
`MiningLevelDeterminent` is a good idea, but a bit of a clunky name. I can't think of a better name in that concept atm, though.

RedstoneParadox replied:
`MiningLevelDeterminent` would imply that it's specific to mining blocks when this interface could (and should) be expanded to other tool stats.

kgscialdone replied:
Perhaps `DynamicToolStats` as suggested in Discord?

LemmaEOF replied:
maybe `DynamicMiningStats` instead, because I'm gonna do a restructure to make this work for any item instead of just tools.

Sollace commented on fabric-mining-levels-v0/src/main/java/net/fabricmc/fabric/mixin/tools/MixinItemStack.java:
```diff
@@ -46,7 +47,13 @@ public void getBlockBreakingSpeed(BlockState state, CallbackInfoReturnable<Float
+				if (this.getItem() instanceof DynamicTool) {
+					miningSpeed = ((DynamicTool) this.getItem()).getMiningSpeed((ItemStack)(Object) this);
+				} else {
+					miningSpeed = ((MiningToolItemAccessor) this.getItem()).getMiningSpeed();
```
Then we have this... What is `MiningToolItemAccessor`?

Should the `DynamicTool` interface be split up? One for mining speed, and another for mining level so you can selectively implement each one independently?

(it's trivial to add a third interface as a union of the two, for convenience's sake)

LemmaEOF replied:
`MiningToolItemAccessor` is a mixin class for accessing the mining speed from `MiningToolItem`, which is protected for some reason. It's not allowed to be implemented by anything other than `MiningToolItem`.

RedstoneParadox replied:
Splitting up the interface would make things more messy when it comes to adding compatibility for other stats such as attack speed and damage.

LemmaEOF replied:
Yeah, it's just not worth it. If you don't want to impl one, you can always just have it pass through to the tool's hardcoded stats.

RedstoneParadox commented on fabric-mining-levels-v0/src/main/java/net/fabricmc/fabric/api/tools/DynamicTool.java:
```diff
@@ -0,0 +1,36 @@
+	 * @param stack The stack to check on.
+	 * @return The mining speed of the tool. 8.0 is equal to a diamond pick.
+	 */
+	float getMiningSpeed(ItemStack stack);
```
`Item` already has a method for getting the mining speed (but strangely not the level).

LemmaEOF replied:
Hm, we might wanna do some rework on how the whole system is set up. Currently we basically skip vanilla behavior for anything that's an instanceof `MiningToolItem` so that we can use our own, non-hardcoded `breakByTool` system, but we might wanna just inject into `MiningToolItem` and change logic there instead. I'll bring it up in the discord and see what people think.

LemmaEOF replied:
Looking at it, the `Item#getMiningSpeed` is what handles speed for effectiveness too, so it's not super cut out for our purposes. The system is currently designed so `ToolManager` is supposed to be internal-only, so we shouldn't set up a scenario where devs have to call ToolManager for effectiveness. There's no way to make sure that all subclasses of MiningToolItem will call to super isEffectiveOn if we use the code there, so it'd be a hassle. We currently implement this stuff in a mixin to ItemStack, which I think is the best option. I'm gonna say we keep the getMiningSpeed option in.

LemmaEOF committed:
rename interface, restructure code

LemmaEOF commented:
Ok, I've made a few changes:
- The interface is now named `DynamicMiningStats`. Feel free to dispute that.
- The code is restructured so that `instanceof DynamicMiningStats` checks are not performed inside of `instanceof MiningToolItemAccessor` if-blocks, allowing *any* item to be given dynamic mining stats, not just mining tools.

I'm thinking about if any other stuff could be done through this. Attack speed and attack damage are done through EntityAttributeModifiers, and those are kind of a mess. I have some ideas of what could be done for them, but it'll be some work.

LemmaEOF commented:
NuclearFarts over in the discord has some [assistant methods](https://pastebin.com/VmZFqUVp) that deal with EAMS by moving them all over into the EAM NBT system Minecraft has for datapack makers. I'm not sure what we want to do with all this, but I think I might have some ideas of how we can make attribute stuff a lot less painful for modders. This PR might end up converting the module into a `fabric-item-attributes-v1` module instead.

LemmaEOF committed:
rename module, add code for dynamic EAMs

LemmaEOF commented:
alright, I finally decided that this is doing enough that it's worth renaming the module. I've added code for ItemStack-based dynamic EntityAttributeModifiers on items, and cleaned up a lot of code. 
- The `MiningToolItemAccessor` is now a general `MixinMiningToolItem` which implements the new-renamed `ToolAttributeHolder`. This removes some of the special-casing that existed previously.
- `ToolAttributeHolder` now has a method for getting dynamic EntityAttributeModifiers, based on the stack and the equipment slot.
- Dynamic EAMs are appended to either a tool's hardcoded EAMs or the EAMs added via NBT tags. Because that might mess with stuff in some cases, adding any tag with the key `"fabric_IgnoreDynamicModifiers"` to an item's NBT will disable appending dynamic EAMs for that item.
- There's a new impl class that does the logic of merging two EAM lists so that they don't give duplicate tooltips. Can be moved if necessary.
- Everything is moved into the new `fabric-tool-attributes-v1` package/mod.

falkreon commented:
Peeling the ToolAttributeHolder interface away from NBT (while providing the facilities to do so if desired) does this api a world of good. Now people have a *lot* of options for reconfiguring items on the fly.

It's too bad that it's not practical to pass a player into these calls. That would open this up even more.

LemmaEOF commented:
I mean, it’s not impossible. Just impractical. If someone figures out a solid use case set that’s worth putting into the interface, then we can think about it more.

Prospector commented:
I don't think you can just rename a module, what if a mod depends on the old one? I think both will have to remain until a proper deprecation cycle has gone by.

LemmaEOF commented:
Oh yeah, that's a fair point. It should probably just stay under fabric-mining-levels-v0 on this branch and change over to fabric-tool-attributes-v1 on the 1.15, maybe.

LemmaEOF commented:
Although actually, for mod-resolution purposes, keeping a mostly-empty fabric-mining-levels-v0 around would work due to the packaging of stuff. There's not really any reason for folks to be JiJ'ing in fabric-mining-levels-v0, so it should be alright as long as the mod is there for MDR.

LemmaEOF committed:
bring back fabric-mining-levels for dep-cycle purposes, fix object-builders dep, license header

LemmaEOF commented:
Alright, fabric-mining-levels-v0 is back with just the deprecated `FabricItemTags`. Everything else is moved to fabric-tool-attributes-v1, and mining-levels depends on tool-attributes. I also changed object-builders to depend on tool-attributes, and added a missing license header in tool-attributes.

LemmaEOF commented:
Actually, I'm gonna verbump mining-levels-v0 so just in case someone *is* JiJ'ing mining-levels-v0, the JiJ'd version will be supplanted by the new mining-levels-v0 that depends on tool-attributes-v1.

LemmaEOF committed:
safety verbump for deprecated mining-levels

JamiesWhiteShirt commented on fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/impl/tools/ToolManager.java:
```diff
@@ -87,7 +88,9 @@ public static void registerBreakByTool(Block block, Tag<Item> tag, int miningLev
-		if (stack.getItem() instanceof ToolItem) {
+		if (stack.getItem() instanceof ToolAttributeHolder) {
+			return ((ToolAttributeHolder) stack.getItem()).getMiningLevel(stack);
+		} else if (stack.getItem() instanceof ToolItem) {
```
ToolItem now implements ToolAttributeHolder, so this special case can be removed.

LemmaEOF replied:
…actually, it doesn't. Only MiningToolItem does. I should fix that.

JamiesWhiteShirt replied:
You forgot to remove this case

JamiesWhiteShirt commented on fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tools/ToolAttributeHolder.java:
```diff
@@ -0,0 +1,52 @@
+	 * @param stack The stack that's equipped.
+	 * @return The dynamic modifiers to add on top of other modifiers on this stack.
+	 */
+	default Multimap<String, EntityAttributeModifier> getDynamicModifiers(EquipmentSlot slot, ItemStack stack) {
```
I don't know how hot this method is. If it is, perhaps we should allow it to return null to avoid allocation of empty Multimaps.

Sollace replied:
Is the returned map modifiable? If not you can have it return a static, immutable multimap with no entries.

LemmaEOF replied:
Multimaps are mutable, and I don't know if there's an immutable multimap. I think it being nullable would be the best option here.

Sollace replied:
What I mean is: Are the maps returned by this method _intended_ to be mutated?

If there isn't an immutable multimap implemented everywhere, I'd be surprised, and I guess I'd agree that `null` would suffice. It's just having a non-null world would make life a lot safer in general.

LemmaEOF replied:
Actually, taking a look at my code again, the returned multimap from `getDynamicModifiers` is never mutated. I'm gonna add an EMPTY immutable multimap for folks to return if they don't want dynamic modifiers.

JamiesWhiteShirt commented on fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/mixin/tools/MixinItemStack.java:
```diff
@@ -42,12 +51,27 @@ public void isEffectiveOn(BlockState state, CallbackInfoReturnable<Boolean> info
+	public void getAttributeModifiers(EquipmentSlot slot, CallbackInfoReturnable<Multimap<String, EntityAttributeModifier>> info, Multimap<String, EntityAttributeModifier> multimap) {
+		ItemStack stack = (ItemStack) (Object) this;
+		if (stack.getItem() instanceof ToolAttributeHolder) {
+			if (!stack.hasTag() || !stack.getTag().containsKey("fabric_IgnoreDynamicModifiers")) {
```
I think we should discuss how we want to approach this. Is it really Fabric's responsibility to add this kind of game logic, or should perhaps implementations of ToolAttributeHolder do it?

Sollace replied:
You can put this logic into the interface as a default method that modders can override if they wish.

LemmaEOF replied:
Yeah, I think I'm gonna add a `useDynamicModifiers` method.

LemmaEOF committed:
switch MiningToolItem mixin to toolItem, clean up EAMs

JamiesWhiteShirt commented on fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tools/ToolAttributeHolder.java:
```diff
@@ -44,9 +46,18 @@
+	 * @return Whether the tool's dynamic modifiers should be appended.
+	 */
+	default boolean useDynamicModifiers(EquipmentSlot slot, ItemStack stack) {
+		return !stack.hasTag() || !stack.getTag().containsKey("fabric_IgnoreDynamicModifiers");
```
I still think this solution is worth discussing. Why should fabric API have this kind of logic?

LemmaEOF replied:
That's a fair point. I put it in because we're technically disrupting vanilla behavior with the dynamic modifiers, and I wanted a way to turn that off. Using NBT tags is what vanilla *always* does in this situation, so I added a tag check in.

LemmaEOF committed:
huh I coulda sworn I got rid of that special-case already

OhiraKyou commented:
> If someone figures out a solid use case set that’s worth putting into the interface, then we can think about it more.

I haven't looked very deeply into the API here. So, some of these are likely out of scope. But, if you're looking for reasons a tool might have to access the player, there are potentially quite a lot. Here are a few examples:
- Tools whose stats scale with health or hunger.
- Parasitic tools that drain health or hunger for durability.
- A dex sword that deals more damage when the player is wearing less armor.
- Tinkers Construct (electrum) style tools that generate static while you move and discharge it for a speed boost.
- Tinkers Construct (pig iron) style edible tools that sacrifice their durability to heal hunger.

RedstoneParadox commented:
> Parasitic tools that drain health or hunger for durability.

Something that should be done during `Item#inventoryTick()`.

> Tinkers Construct (electrum) style tools that generate static while you move and discharge it for a speed boost.

Again, something that should be done during `Item#inventoryTick()`.

> Tinkers Construct (pig iron) style edible tools that sacrifice their durability to heal hunger.

Which should be taken care of in `Item#use`.

The other two probably do need player access inside the interface methods.

LemmaEOF commented:
Yeah, we discussed passing a player into calls but it's not really that practical. We can't guarantee there's always a player around. We could add a nullable player, but I'm just not sure at the moment.

OhiraKyou commented:
Well, it wouldn't necessarily need to be a player. It's common practice to pass an instigator object when dealing damage (or sender, for general events). So, it wouldn't be unusual for a tool to have a "user" of some description. That could be a block, player, or any other entity. It could even be itself, if it's a sentient tool.

LemmaEOF commented:
Hmm, I'll think about it.

LemmaEOF committed:
make useDynamicModifiers default to true instead

Boundarybreaker commented:
Yeah, I don't really think there's any good way to add actors without *massively* complicating the system. We'd probably need mixins into three or four more places, and it just doesn't seem reasonable. If someone finds a way to do it, that can be submitted later. I *am* pushing forward a change so that there's no default way to disable dynamic modifiers on a tool.

LemmaEOF commented:
Yeah, I don't really think there's any good way to add actors without *massively* complicating the system, since we primarily mix into ItemStack and that doesn't have a way of knowing what inventory it's in or who it's being used by. We'd probably need mixins into three or four more places, and it just doesn't seem reasonable. If someone finds a way to do it, that can be submitted later. I *am* pushing forward a change so that there's no default way to disable dynamic modifiers on a tool.

LemmaEOF commented:
Actually, lemme take another look. I might have some ideas.

LemmaEOF committed:
add actor-based attribute getting

LemmaEOF commented:
I actually managed to implement it! However, that comes at the cost of needing to use some `@Redirect`s in `MixinLivingEntity`, so that the passed stack can have its actor-based EAMs properly added. I'm not sure if that's kosher, and there might be a better way to go about this, but it doesn't seem like you can capture locals with a `@ModifyArg`, so getting the stack and slot wouldn't be doable with that. I feel like can't just inject before or after the standard `removeAll()` or `replaceAll()` calls, because that could lead to EAMs overriding each other. If someone else knows how to do this better, please let me know! Worse comes to worst, we can just not support EAMs on actor tools, but that'd leave a bit of a sour taste in my mouth.

LemmaEOF committed:
move to v1 package, start cleaning up whitespace

liach commented on fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tools/v1/ToolActor.java:
```diff
@@ -0,0 +1,17 @@
+package net.fabricmc.fabric.api.tools.v1;
+
+import com.sun.istack.internal.Nullable;
```
:thinking:

LemmaEOF replied:
I *swore* that I imported javax Nullable.

liach commented on fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/impl/tools/AttributeManager.java:
```diff
@@ -0,0 +1,74 @@
+import java.util.HashSet;
+import java.util.Set;
+
+public class AttributeManager {
```
final and private constructor

liach commented on fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tools/v1/ToolAttributeHolder.java:
```diff
@@ -0,0 +1,63 @@
+	 * @param stack The stack that's equipped.
+	 * @return Whether the tool's dynamic modifiers should be appended.
+	 */
+	default boolean useDynamicModifiers(EquipmentSlot slot, ItemStack stack) {
```
A conditional return for `getDynamicModifiers` should suffice. This method is extraneous.

liach commented on fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/impl/tools/EntityToolActor.java:
```diff
@@ -0,0 +1,17 @@
+import net.fabricmc.fabric.api.tools.v1.ToolActor;
+import net.minecraft.entity.LivingEntity;
+
+public class EntityToolActor implements ToolActor<LivingEntity> {
```
Make it a generic simple tool actor (like `BasicToolActor<T>`), or even better, make it a package-private class at the same package as `ToolActor`, and add a simple static method in `ToolActor` like `static <T> ToolActor<T> of(T actor)`

liach commented on fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tools/v1/ToolActor.java:
```diff
@@ -0,0 +1,17 @@
+ * @param <T> The type of actor that will use this tool.
+ */
+public interface ToolActor<T> {
+	ToolActor NO_ACTOR = () -> null;
```
generics.

LemmaEOF replied:
Not sure how to do generics if there's a null.

liach replied:
`java.lang.Void` like in `CompletableFuture<Void>`. I suggest moving this field to another package private class and use a method to get the dummy actor instead (like java `Collections.emptySet()` (type safe) vs `Collections.EMPTY_SET` (not type safe))

liach commented on fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tools/v1/ToolActor.java:
```diff
@@ -0,0 +1,17 @@
+	 * @return The actor using a tool, if they exist.
+	 */
+	@Nullable
+	T getActor();
```
`actor.getActor()` sounds weird to me.

liach commented on fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/impl/tools/ToolManager.java:
```diff
@@ -86,28 +88,34 @@ public static void registerBreakByTool(Block block, Tag<Item> tag, int miningLev
-		if (stack.getItem() instanceof ToolItem) {
-			return ((ToolItem) stack.getItem()).getMaterial().getMiningLevel();
+	private static int getMiningLevel(ItemStack stack, ToolActor actor) {
+		if (stack.getItem() instanceof ActableAttributeHolder && !actor.equals(ToolActor.NO_ACTOR)) {
```
I suggest checking whether the actor's actual actor is null instead of comparing identity.

liach commented on fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/impl/tools/ToolManager.java:
```diff
@@ -86,28 +88,34 @@ public static void registerBreakByTool(Block block, Tag<Item> tag, int miningLev
+	 * Use {@link ToolActor#NO_ACTOR} if there's not an applicable actor for this scenario (i.e, you only have the item stack)
 	 */
-	public static TriState handleIsEffectiveOn(ItemStack stack, BlockState state) {
+	public static TriState handleIsEffectiveOn(ItemStack stack, BlockState state, ToolActor actor) {
```
generics for ToolActor

liach commented on settings.gradle:
```diff
@@ -24,7 +24,6 @@ include 'fabric-events-lifecycle-v0'
 include 'fabric-item-groups-v0'
 include 'fabric-keybindings-v0'
 include 'fabric-loot-tables-v1'
-include 'fabric-mining-levels-v0'
```
Why don't you remove the module directory wholesale if you remove it from gradle build?

LemmaEOF replied:
Oh, shit. It still needs to be included for deprecation purposes.

liach commented:
Also consider fixing merge conflict for mod.json

Commit:
Merge branch '1.14' into patch-3

LemmaEOF committed:
implement feedback

liach commented on fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tools/v1/ActableAttributeHolder.java:
```diff
@@ -0,0 +1,41 @@
+	 * @param actor The actor attempting to mine.
+	 * @return The mining level for this stack and actor.
+	 */
+	default int getMiningLevel(ItemStack stack, ToolActor actor) {
```
Rawtypes

liach commented on fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/impl/tools/AttributeManager.java:
```diff
@@ -0,0 +1,76 @@
+	 * @return The merged form of the two.
+	 */
+	public static Multimap<String, EntityAttributeModifier> mergeAttributes(Multimap<String, EntityAttributeModifier> left, Multimap<String, EntityAttributeModifier> right) {
+		Multimap<String, EntityAttributeModifier> ret = HashMultimap.create();
```
Consider optimizing when left or right is empty

liach commented on fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/mixin/tools/MixinLivingEntity.java:
```diff
@@ -0,0 +1,55 @@
+@Mixin(LivingEntity.class)
+public abstract class MixinLivingEntity extends Entity {
+
+	private ToolActor actor = ToolActor.of((LivingEntity)(Object)this);
```
rawtypes

liach commented on fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/mixin/tools/MixinPlayerInventory.java:
```diff
@@ -0,0 +1,61 @@
+
+@Mixin(PlayerInventory.class)
+public abstract class MixinPlayerInventory {
+	private final ToolActor actor = ToolActor.of(this.player);
```
rawtypes

LemmaEOF committed:
more clean-up

ghost commented:
Some java files are missing license headers? 
- fabric-mining-levels-v0/FabricToolTags
- fabric-tool-attributes-v1/ActableAttributeHolder
- fabric-tool-attributes-v1/ToolActor
- fabric-tool-attributes-v1/MixinLivingEntity
- fabric-tool-attributes-v1/MixinPlayerInventory

edit: looks good though

LemmaEOF commented:
ah piss license headers

LemmaEOF committed:
license headers!

LemmaEOF commented:
This has been lying around for a while. Anyone else have feedback, or should it head to last call?

shedaniel commented:
Might want to fix the merge conflicts introduced by the formatting code style commit.

Player3324 commented:
Please rebase and try to build it for checkstyle.

LemmaEOF committed:
Merge remote-tracking branch 'upstream/1.14' into patch-3

LemmaEOF committed:
fix some checkstyle violations

LemmaEOF committed:
more reformatting

LemmaEOF commented:
@sfPlayer1 I've done all I can do to fix checkstyles, but it seems like our checkstyle is broken. I've got [this set of lines](https://github.com/FabricMC/fabric/blob/fea4325282347326f7150ed2669b558e6051384c/fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/impl/tools/AttributeManager.java#L28-L39) in AttributeManager that, no matter how I try to fix them, refuses to pass checkstyle due to conflicting rules about whitespace. It seems like the only solution is to get rid of the javadocs, but I am not willing to do that.

LemmaEOF commented:
Okay, for some reason removing the tab fixies it. Weird.

LemmaEOF committed:
fix checkstyle issue

Player3324 commented:
This was most likely matched by the "trailing whitespace" pattern, which appears to have worked as expected. A blank line shouldn't have anything in it, most editors have a setting to strip it automatically on save.

LemmaEOF committed:
repackage things

LemmaEOF commented:
moved from `net.fabricmc.fabric.tools` to `net.fabricmc.fabric.tool.attribute` as player requested in discord

liach commented:
Mind fixing merge conflicts as well?

LemmaEOF commented:
I'm trying, but git *really* doesn't want to let me.

LemmaEOF commented:
Superseded on 1.15 by #460.
